test/web-platform-tests/run-single-wpt.js

```diff
@@ -177,28 +192,58 @@ function createJSDOM(urlPrefix, testPath, expectFail) {
               window.close();
             });
 
+            let harnessFail = false;
             if (harnessStatus.status === harnessStatus.ERROR) {
+              harnessFail = true;
               errors.push(new Error(`test harness should not error: ${testPath}\n${harnessStatus.message}`));
             } else if (harnessStatus.status === harnessStatus.TIMEOUT) {
+              harnessFail = true;
               errors.push(new Error(`test harness should not timeout: ${testPath}`));
             }
 
-            errors.push(...doneErrors);
             errors.push(...unhandledExceptions);
 
+            if (typeof expectFail === "object" && (harnessFail || unhandledExceptions.length)) {
+              expectFail = false;
+            }
+
             if (errors.length === 0 && expectFail) {
-              reject(new Error(`
-            Hey, did you fix a bug? This test used to be failing, but during
-            this run there were no errors. If you have fixed the issue covered
-            by this test, you can edit the "to-run.yaml" file and remove the line
-            containing this test. Thanks!
-            `));
-            } else if (errors.length === 1 && !expectFail) {
+              reject(new Error(unexpectedPassingTestMessage));
+            } else if ((errors.length === 1 && (tests.length === 1 || harnessFail)) && !expectFail) {
               reject(new Error(errors[0]));
             } else if (errors.length && !expectFail) {
-              reject(new Error(`${errors.length} errors in test:\n\n${errors.join("\n")}`));
-            } else {
+              reject(new Error(`${errors.length}/${tests.length} errors in test:\n\n${errors.join("\n\n")}`));
+            } else if (typeof expectFail !== "object") {
               resolve();
+            } else {
+              const unexpectedErrors = [];
+              for (const test of tests) {
+                const testFailed =
+                  test.status === test.FAIL ||
+                  test.status === test.TIMEOUT ||
+                  test.status === test.NOTRUN;
+                const data = expectFail[test.name];
+                let reason;
+
+                if (Array.isArray(data)) {
```

Comment:
Won't it always be an array?

Reply:
Not in the new format, where it will be a plain object.

Reply:
Only if there’s a `to‑run.yaml` entry for this subtest, otherwise it will be `undefined`.